arango/connections/base.py

```diff
@@ -0,0 +1,259 @@
+from __future__ import absolute_import, unicode_literals
+
+import logging
+
+from arango.jobs import BaseJob, OldJob
+from arango.http_clients import DefaultHTTPClient
+from arango.utils import sanitize, fix_params
+from arango.api.wrappers import AQL, Graph
+from arango.api.collections import Collection
+
+
+class BaseConnection(object):
+    """ArangoDB database connection.
+
+    :param protocol: the internet transfer protocol (default: ``"http"``)
+    :type protocol: str | unicode
+    :param host: ArangoDB host (default: ``"localhost"``)
+    :type host: str | unicode
+    :param port: ArangoDB port (default: ``8529``)
+    :type port: int | str | unicode
+    :param database: the name of the target database (default: ``"_system"``)
+    :type database: str | unicode
+    :param username: ArangoDB username (default: ``"root"``)
+    :type username: str | unicode
+    :param password: ArangoDB password (default: ``""``)
+    :type password: str | unicode
+    :param http_client: the HTTP client
+    :type http_client: arango.clients.base.BaseHTTPClient
+    :param enable_logging: log all API requests with a logger named "arango"
+    :type enable_logging: bool
+    """
+
+    def __init__(self,
+                 protocol='http',
+                 host='localhost',
+                 port=8529,
+                 database="_system",
+                 username='root',
+                 password='',
+                 http_client=None,
+                 enable_logging=True,
+                 logger=None,
+                 old_behavior=True):
+
+        if http_client is None:
+            http_client = DefaultHTTPClient()
+
+        if logger is None:
+            logger = logging.getLogger('arango')
+
+        self._protocol = protocol.strip('/')
+        self._host = host.strip('/')
+        self._port = port
+        self._database = database
+        self._url_prefix = '{protocol}://{host}:{port}/_db/{db}'.format(
+            protocol=self._protocol,
+            host=self._host,
+            port=self._port,
+            db=self._database
+        )
+        self._username = username
+        self._password = password
+        self._http = http_client
+        self._enable_logging = enable_logging
+        self._type = 'standard'
+        self._logger = logger
+        self._old_behavior = old_behavior
+
+        if old_behavior:
+            self._default_job = OldJob
```

Comment:
So here's the context for the names.  It isn't accurate to describe BaseJob as AsyncJob- there is nothing inherently asynchronous about its behavior.  All it does is lazy loading of variables to prevent accessing the response until result is called.  This is a requirement for asynchronous dispatch, but is not asynchronous in and of itself.  The difference between the two is that OldJob is a mask which has the same interface as the rest of the Jobs to enable using the class in the same manner, but instead of an instance of BaseJob, it returns the result of BaseJob.result(raise_errors=True) to maintain the same behavior as previously.  OldJob may not be the best name, but I don't think SyncJob is good either.